### PR TITLE
IDE-729 Auto Complete Sprint 1

### PR DESCRIPTION
**Pull request by dehilsterlexis**
Fixed IDE-729

Signed-off-by: dehilsterlexis <david.dehilster@lexisnexis.com>

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/EclMeta.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#include "StdAfx.h"
+
+#include "EclMeta.h"
+
+CEclMeta::CEclMeta()
+{
+}
+
+void CEclMeta::LoadMetaData(IEclCC *eclcc)
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: CEclMeta should not "know" about IEclCC.  You could simply pass in an array of the root folders here instead.

File: comms/EclMeta.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+{
+    if (eclcc)
+    {
+        for (int i = 0; i < eclcc->GetEclFolderCount(); ++i)
+        {
+            CEclMeta::PopulateMeta(boost::filesystem::wpath(eclcc->GetEclFolder(i), boost::filesystem::native),_T(""));
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: remove "CEclMeta" prefix (doesn't this imply your calling a static method?

Comment (GordonSmith) [reply comms/EclMeta.cpp]: I think there is a stringToWPath helper?

File: wlib/SourceView.cpp [hunk 3]
```diff
@@ -150,6 +151,13 @@ void CSourceCtrl::SetSourceType(const CString & typeStr)
     }
 }
 
+void CSourceCtrl::SetAttribute(IAttribute * attr)
+{
+    ATLASSERT(attr);
+    m_attribute = attr;
+    SetType(m_attribute->GetType());
+}
```
Comment (GordonSmith) [inline wlib/SourceView.cpp]: Was there not some sort of "SetAttribute" already?

File: wlib/SourceView.cpp [hunk 4]
```diff
@@ -919,7 +932,18 @@ bool CSourceCtrl::StartAutoComplete()
         CString module(word, periodPos);
         if (m_type->IsTypeOf(ATTRIBUTE_TYPE_ECL))
         {
-            m_langRef->GetAutoC((const TCHAR *)module, list);
+            if (CComPtr<IEclCC> eclcc = CreateIEclCC())
+            {
+                if (boost::algorithm::starts_with(static_cast<const TCHAR *>(word),_T("$")))
```
Comment (GordonSmith) [inline wlib/SourceView.cpp]: This logic should be inside the Meta Class as it "knows" more about ECL than the editor will.

File: comms/EclMeta.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#include <global.h>
+#include "EclCC.h"
+#include <UtilFilesystem.h>
+
+enum META_TYPE
+{
+    META_TYPE_UNKNOWN = -1,
+    META_TYPE_FOLDER,
+    META_TYPE_FILE
+};
```
Comment (GordonSmith) [inline comms/EclMeta.h]: If you get  your overloading methods done right, then this will not be needed.

File: comms/EclMeta.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+typedef std::map<std::wstring, CEclMetaData> EclMetaDataMap;
+
+class CEclMeta
+{
+public:
+    EclMetaDataMap m_masterMeta;
+
+    CEclMeta::CEclMeta();
+    void CEclMeta::LoadMetaData(IEclCC *eclcc);
+    void CEclMeta::PopulateMeta(const boost::filesystem::wpath & fileOrDir, std::wstring path, int level=0);
```
Comment (GordonSmith) [inline comms/EclMeta.h]: Remove "CEclMeta::"

File: wlib/SourceView.h [hunk 7]
```diff
@@ -123,6 +123,7 @@ class CSourceCtrl : public CWindowImpl<CSourceCtrl, CScintillaCtrlEx, CScintilla
     bool m_modified;
     StlLinked<ILangRef> m_langRef;
     CComPtr<IAttributeType> m_type;
+    IAttribute *m_attribute;
```
Comment (GordonSmith) [inline wlib/SourceView.h]: Remove - and add a new method to ISourceSlot "GetAutoCList or similar".

File: wlib/SourceView.cpp [hunk 8]
```diff
@@ -509,7 +517,12 @@ void CSourceCtrl::InitColors(ILangRef * langRef)
     AnnotationSetVisible(ANNOTATION_BOXED);
 }
 
-bool CSourceCtrl::IsWordCharForSelWithPreiod(char ch) 
+bool CSourceCtrl::IsWordCharForSelWithPeriodECL(char ch) 
+{
+    return (strchr(wordchars, ch) != NULL || ch == '.' || ch == '$');
```
Comment (GordonSmith) [inline wlib/SourceView.cpp]: `return IsWordCharForSelWithPeriod(ch) || ch =='$'`

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/EclMeta.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            if (dottedPath.empty())
+                newDottedPath = fileOrDir.stem().c_str();
+            else
+                newDottedPath = dottedPath + _T(".") + fileOrDir.stem().c_str();
+        }
+
+        if (HasValidExtension(fullpath))
+        {
+            m_masterMeta[newDottedPath] = *(new CEclFile(fileOrDir));
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: Memory leak

File: comms/EclMeta.cpp [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        else if (!fileOrDir.has_extension())
+        {
+            if (level)
+                m_masterMeta[newDottedPath] = *(new CEclFolder(fileOrDir));
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: Memory leak

**Comment by dehilsterlexis** [comment]
@richardkchapman I have been working on the changes.

**Comment by richardkchapman** [comment]
@dehilsterlexis

File: comms/EclCC.cpp [hunk 11]
```diff
@@ -49,6 +47,7 @@ class CEclCC : public IEclCC, public clib::CLockableUnknown
 public:
     std::_tstring m_errors;
     std::_tstring m_warnings;
+    mutable CEclMeta m_eclMeta;
```
Comment (GordonSmith) [inline comms/EclCC.cpp]: Probably doesn't need to be mutable?

File: comms/EclMeta.h [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+class CEclMetaData : public clib::CLockableUnknown
+{
+protected:
+    boost::filesystem::wpath m_path;
+    std::_tstring m_moduleLabel;
+public:
+    BEGIN_CUNKNOWN
+    END_CUNKNOWN(CUnknown)
+
+    boost::filesystem::wpath GetPath() { return m_path; }
```
Comment (GordonSmith) [inline comms/EclMeta.h]: should be const

File: comms/EclMeta.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    bool HasPath() { return m_path.size() > 0; }
```
Comment (GordonSmith) [inline comms/EclMeta.h]: should be const

File: comms/EclMeta.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+class CEclMeta
+{
+public:
+    EclMetaDataMap m_masterMeta;
+
+    void LoadMetaData(WPathVector *folders);
+    void PopulateMeta(const boost::filesystem::wpath & fileOrDir, std::wstring path, int level=0);
+    StlLinked<CEclMetaData> GetMetaInfo(const std::_tstring & moduleFile);
+    bool AddAutoStr(StdStringVector &set, const std::_tstring & str);
+    bool GetAutoC(const std::_tstring & partialLabel, StdStringVector &set);
```
Comment (GordonSmith) [inline comms/EclMeta.h]: should be const

File: comms/EclMeta.h [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+typedef std::map<std::wstring, StlLinked<CEclMetaData> > EclMetaDataMap;
+
+class CEclMeta
+{
+public:
+    EclMetaDataMap m_masterMeta;
+
+    void LoadMetaData(WPathVector *folders);
+    void PopulateMeta(const boost::filesystem::wpath & fileOrDir, std::wstring path, int level=0);
+    StlLinked<CEclMetaData> GetMetaInfo(const std::_tstring & moduleFile);
```
Comment (GordonSmith) [inline comms/EclMeta.h]: should be const

File: eclide/ChildAttributeFrame.cpp [hunk 16]
```diff
@@ -42,6 +42,8 @@ class CAttributeFrame : public CChildFrame, public boost::signals::trackable,
     BEGIN_CLOCKABLEUNKNOWN
     END_CUNKNOWN(CChildFrame)
 
+    void GetCompletionList(const std::_tstring & module, StdStringVector &list) {}
```
Comment (GordonSmith) [inline eclide/ChildAttributeFrame.cpp]: should be const

File: comms/EclMeta.cpp [hunk 17]
```diff
@@ -0,0 +1,74 @@
+#include "StdAfx.h"
+
+#include "EclMeta.h"
+
+void CEclMeta::LoadMetaData(WPathVector *folders)
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: Should be `const WPathVector & folders`

File: comms/EclMeta.cpp [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+{
+    for (WPathVector::const_iterator itr = folders->begin(); itr != folders->end(); ++itr)
+    {
+        PopulateMeta(itr->first,_T(""));
+    }
+}
+
+void CEclMeta::PopulateMeta(const boost::filesystem::wpath & fileOrDir, std::wstring dottedPath, int level)
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: Should be `const std::wstring & dottedPath`

File: comms/EclCC.cpp (hunk 11 above)
Comment (dehilsterlexis) [reply comms/EclCC.cpp]: @GordonSmith it turns out that it doesn't need to be mutable if GetAutoC is changed to a non-constant. Going that route.

**Comment by dehilsterlexis** [comment]
@GordonSmith back to you.

**Comment by richardkchapman** [comment]
Closing as per discussion with Gordon

**Pull request closed without merging**